src/cluster_asm.c

```diff
@@ -1346,11 +1644,74 @@ void asmBeforeSleep(void) {
             asmStartSyncSlots(task);
         } else if (task->state == ASM_STREAMING_BUF) {
             asmSyncBufferStreamToDb(task);
+        }
+    }
+
+    if (task->operation == ASM_MIGRATE) {
+        if (task->state == ASM_PAUSED_WRITE) {
+            client *c = task->main_channel_client;
+            /* The command streams for slot ranges have been drained. */
+            if (!clientHasPendingReplies(c)) {
+                serverLog(LL_NOTICE, "The command streams for slot ranges have been drained, sending STREAM-EOF");
+
+                if (unlikely(asmIsFailPointActive(ASM_MIGRATE_MAIN_CHANNEL, task->state)))
+                    connShutdown(c->conn);
+
+                /* Send STREAM EOF, must use this approach to guarantee this command delivery */
+                char *err = sendCommand(c->conn, "CLUSTER", "SYNCSLOTS", "STREAM-EOF", NULL);
+                if (err) {
+                    serverLog(LL_WARNING, "Error sending CLUSTER SYNCSLOTS STREAM-EOF: %s", err);
+                    asmTaskSetError(task, "Migrate main channel sending STREAM-EOF failed: %s, state: %s",
+                                          err, asmTaskStateToString(task->state));
+                    asmMigrateFailed(task);
+                    sdsfree(err);
+                    return;
+                }
+
+               /* Even though the main channel client is no longer needed, we can't
+                * close it directly because the destination may still be sending ACKs
+                * over this connection. Instead, we leave it to the destination to
+                * close it. We just clear the task and client references */
+                task->main_channel_client->task = NULL;
+                task->main_channel_client = NULL;
+
+                task->state = ASM_STREAM_DONE;
+                /* Notify plugin import is completed */
+                clusterAsmOnEvent(task->slot_ranges, ASM_EVENT_MIGRATE_WAIT_FINALIZE, NULL);
+            }
         } else if (task->state == ASM_FAILED) {
-            asmTaskReset(task);
-            task->retry_count++;
-            serverAssert(task->state == ASM_NONE);
-            asmStartSyncSlots(task);
+            task->done_time = server.mstime; /* Don't retry */
+
+            /* TODO: if it is necessary to save this task? As the destination side will retry,
+             * there may be a lots of failed tasks. */
+            /* Move task to done list, for migrate task, the source side doesn't retry,
+             * the destination side does or cancels. */
```

Comment:
perhaps we should try to match failed tasks and new attempts. If slotrange and source id matches, it is another retry. So, we reuse same task and won't generate bunch of new ones on the source side.

Reply:
good idea